Helm chart

bikash119 opened the pull request:
# Description
<!-- Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change. -->
- Use helm chart to manage k8s resources
- Added pre-commit hook to use `helm lint` to find and fix linting errors in chart templates.

Closes #<issue_number>

**Type of change**
<!--  Please delete options that are not relevant. Remember to title the PR according to the type of change  -->

- Improvement (change adding some improvement to an existing functionality)
- Documentation update

**How Has This Been Tested**
<!--  Please add some reference about how your feature has been tested.  -->
on dev machine
- install minikube
- install helm
- helm install <name> <chart_folder_location> . In this case `helm install argilla-dev examples/deployment/k8s/argilla-cart`
This should install and start elasticsearch and argilla-server to the k8s cluster
- `kubectl port-forward service/argilla-dev-argilla-chart 6900:6900`. 
- http://localhost:6900 ->should open the Argilla UI login screen.
- USERNAME: argilla , PASSWORD: 12345678, will authenticate the user and take user to http://localhost:6900/datasets.

**Checklist**
<!--  Please go over the list and make sure you've taken everything into account -->

- I added relevant documentation [TODO]
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation [TODO]
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)[ TODO ]

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

Arvind644 committed:
added helm chart

Arvind644 committed:
fixed indentation in Chart.yaml

Arvind644 committed:
fixes helm chart

Arvind644 committed:
Merge branch 'develop' of https://github.com/Arvind644/argilla into develop

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

Arvind644 committed:
fixed helm chart readme

bikash119 committed:
Merge remote-tracking branch 'upstream/develop' into helm_chart

bikash119 committed:
First working commit towards helm chart based k8s resource creation.

bikash119 committed:
Working copy of making elasticsearch either a depedency chart or using external elasticsearch

bikash119 committed:
Fixed linting issues with k8s resource yamls

bikash119 committed:
Merge remote-tracking branch 'upstream/develop' into helm_chart

bikash119 committed:
Made configmap optional based on if elasticsearch is installed or used as external service

bikash119 committed:
added redis as a dependency for Argilla chart

bikash119 committed:
Elasticsearch operator used for deployment of elasticsearch

bikash119 committed:
removed un-necessary file which was used to test elasticsearch deployment

bikash119 committed:
- Introduced PVC to store argilla data
- Removed k8s resource yamls

bikash119 committed:
Removed k8s folder in the under root argilla folder

bikash119 committed:
- Elasticsearch deployment is made optional.
- Elasticsearch "built in" is deployed as operator.
- Redis, Elasticsearch, Argilla user PV to persist data beyond restarts.
- Argilla owner username and password is passed from values.yaml.
- Align HPA config with “conventions” – I.e. enable via flag and have static replicaCount variable if not activated.
- Allow for injecting some configuration via configmaps / secrets etc. If external ES is used, the entries to /etc/hosts are pushed using cm.
- Ingress class is made configurable and can be set in values.yaml.
- TODO: Add a readme.md for developers to use it.
- TODO: Add unit tests to check persistence storage usage.

bikash119 committed:
A getting started guide

bikash119 committed:
Modified the output of Notes.txt to increase readability.

bikash119 committed:
First commit : helm unit test [ WIP ]

bikash119 committed:
All the unit testcases are passing

bikash119 committed:
- Updated README.md to add instruction to execute helm unittest.
 - Updated CHANGELOG.md to include the PR for helm chart based installation.

bikash119 committed:
fixed typo error

bikash119 committed:
fixed typo error

bikash119 committed:
Merge branch 'develop' into helm_chart

bikash119 committed:
Merge branch 'develop' into helm_chart

frascuchon commented on examples/deployments/k8s/argilla-chart/README.md:
```diff
@@ -0,0 +1,104 @@
+
+- Redis
+- Elasticsearch
+## Adding the Helm Repository
```
```suggestion
- Elasticsearch

## Adding the Helm Repository
```

frascuchon commented on argilla-server/CHANGELOG.md:
```diff
@@ -15,7 +15,8 @@ These are the section headers that we use:
 ## [Unreleased]()
-
+### Added
+- Added helm chart for argilla. ([#5512](https://github.com/argilla-io/argilla/pull/5512))
```
```suggestion

### Added

- Added helm chart for argilla. ([#5512](https://github.com/argilla-io/argilla/pull/5512))

```

frascuchon commented on examples/deployments/k8s/argilla-chart/README.md:
````diff
@@ -0,0 +1,104 @@
+```bash
+helm repo update
+helm upgrade my-release argilla/argilla
+```
````
```suggestion
After adding the repository, you can install the chart with the release name `my-argilla-server`:

```bash
helm install my-argilla-server examples/deployments/k8s/argilla-chart
```

## Configuration

To override default values, override the values in `values.yaml` file and install the chart using:

```bash
helm install my-argilla-server examples/deployments/k8s/argilla-chart
```

## Upgrading

To upgrade the chart to a newer version:

```bash
helm repo update
helm upgrade my-argilla-server argilla/argilla
```
```

frascuchon commented on examples/deployments/k8s/argilla-chart/README.md:
```diff
@@ -0,0 +1,104 @@
+
+## Uninstalling the Chart
+
+To uninstall/delete the `my-release` deployment:
```
```suggestion
To uninstall/delete the `my-argilla-server` deployment:
```

frascuchon commented on examples/deployments/k8s/argilla-chart/README.md:
````diff
@@ -0,0 +1,104 @@
+To uninstall/delete the `my-release` deployment:
+
+```bash
+helm delete my-release
````
```suggestion
helm delete my-argilla-server
```

frascuchon commented on examples/deployments/k8s/argilla-chart/templates/NOTES.txt:
```diff
@@ -0,0 +1,54 @@
+  {{ .Release.Name }}-argilla-pvc
+{{- end }}
+
+{{- if .Values.elasticsearch.persistence.enabled }}
```
This should be also conditioned to the useOperator flag, right?

bikash119 replied:
Thank you @frascuchon for reviewing my work. I have incorporated these suggestions.

frascuchon commented on examples/deployments/k8s/argilla-chart/templates/NOTES.txt:
```diff
@@ -0,0 +1,54 @@
+{{- else }}
+Using external Redis at:
+  {{ .Values.externalRedis.host }}:{{ .Values.externalRedis.port }}
+{{- end }}
```
Swapping redis notifications: first installation, then persistence storage.

```suggestion
{{- if .Values.redis.enabled }}
Redis is running as a dependency of this chart. You can access it at:
  {{ .Release.Name }}-redis-master:6379
{{- else }}
Using external Redis at:
  {{ .Values.externalRedis.host }}:{{ .Values.externalRedis.port }}
{{- end }}

{{- if .Values.redis.master.persistence.enabled }}
Persistence for Redis is enabled. Your Redis data will be stored in a Persistent Volume Claim named:
  {{ .Release.Name }}-redis-pvc
{{- end }}
{{- if and (not .Values.redis.master.persistence.enabled) (not .Values.elasticsearch.persistence.enabled) (not .Values.argilla.persistence.enabled)}}
Persistence is disabled. Your data will not be preserved across pod restarts.
{{- end }}
```

frascuchon commented on examples/deployments/k8s/argilla-chart/templates/NOTES.txt:
```diff
@@ -0,0 +1,54 @@
+  {{ .Release.Name }}-elasticsearch-pvc
+{{- end }}
+
+{{- if .Values.redis.master.persistence.enabled }}
```
This should be conditioned also to redis.enabled flag, right?

bikash119 replied:
Thank you @frascuchon for reviewing my work. I have incorporated these suggestions.

frascuchon commented on examples/deployments/k8s/argilla-chart/values.yaml:
```diff
@@ -0,0 +1,81 @@
+    className: "nginx"
+    host: "argilla.local"
+  configmap:
+    minikubeIP: "192.168.64.2"
```
Are these a values example for a specific environment?

bikash119 replied:
To test locally, I used Minikube as the Kubernetes cluster. To simulate an external Elasticsearch (ES), I installed ES and accessed it via ingress. I used a ConfigMap to add entries to `/etc/hosts`, allowing Argilla to resolve the external ES host.
ConfigMap resource yaml 
```
{{- if not .Values.elasticsearch.useOperator }}
apiVersion: v1
kind: ConfigMap
metadata:
  name: custom-hosts
data:

  hosts: |
    127.0.0.1 localhost
    ::1 localhost
    {{ .Values.argilla.configmap.minikubeIP }} {{ .Values.argilla.ingress.host }}
{{- end }}
```

frascuchon reviewed (commented):
Thanks @bikash119  for this amazing PR.

I let some comments/questions. I will try the installation later.

bikash119 committed:
Update examples/deployments/k8s/argilla-chart/README.md

Accept change suggested by @frascuchon

Co-authored-by: Paco Aranda <francisco.aranda@huggingface.co>

bikash119 committed:
Update argilla-server/CHANGELOG.md

Accept change suggested by @frascuchon

Co-authored-by: Paco Aranda <francisco.aranda@huggingface.co>

bikash119 committed:
Update examples/deployments/k8s/argilla-chart/README.md

Accept change suggested by @frascuchon

Co-authored-by: Paco Aranda <francisco.aranda@huggingface.co>

bikash119 committed:
Update examples/deployments/k8s/argilla-chart/README.md

Accept change suggested by @frascuchon

Co-authored-by: Paco Aranda <francisco.aranda@huggingface.co>

bikash119 committed:
Update examples/deployments/k8s/argilla-chart/README.md

Accept change suggested by @frascuchon

Co-authored-by: Paco Aranda <francisco.aranda@huggingface.co>

bikash119 committed:
Update examples/deployments/k8s/argilla-chart/templates/NOTES.txt

Accept change suggested by @frascuchon

Co-authored-by: Paco Aranda <francisco.aranda@huggingface.co>

bikash119 committed:
- Added steps to install ECK Operator.
 - Added steps to get started in minikube cluster for local development

bikash119 committed:
- Added readiness and liveliness probe to elasticsearch
- Updated README.md to include few missing steps

bikash119 committed:
Merge remote-tracking branch 'upstream/develop' into helm_chart

bikash119 committed:
Merge branch 'develop' into helm_chart

bikash119 committed:
- Updated README.md
- added a resource yaml for argilla-worker
- fixed the ports used for readiness and liveliness probes

bikash119 committed:
- Added api-key to enable integration test
- Added labels and selector-labels for worker

bikash119 committed:
Updated the README to include steps to check status of deployement and run integration tests

bikash119 committed:
- fixed typo

bikash119 committed:
- fixed typo

bikash119 committed:
- few more instruction to check if all the intermediate steps complete successfully before running integration test

bikash119 committed:
Merge branch 'develop' into helm_chart

bikash119 committed:
Merge branch 'develop' into helm_chart

bikash119 committed:
Merge branch 'develop' into helm_chart

bikash119 committed:
Merge branch 'develop' into helm_chart

frascuchon reviewed (approved):
Thanks for the contribution @bikash119!!!

bikash119 committed:
Merge branch 'develop' into helm_chart
